main/src/ca/uwaterloo/flix/language/phase/unification/BoolUnification.scala

```diff
@@ -75,37 +76,69 @@ object BoolUnification {
       case _ => // nop
     }
 
-    // translate the types into formulas
-    implicit val alg: BoolAlg[BoolFormula] = BoolFormula.AsBoolAlg
-
-    val env = alg.getEnv(List(tpe1, tpe2))
-    val f1 = alg.fromType(tpe1, env)
-    val f2 = alg.fromType(tpe2, env)
-
-    val renv = alg.liftRigidityEnv(renv0, env)
+    val typeVars = tpe1.typeVars ++ tpe2.typeVars
+    val varThreshold = 4
+    if(typeVars.size > varThreshold) {
+      implicit val alg: BoolAlg[DD] = BddFormula.AsBoolAlg
```

Comment:
It occurs to me this code is now duplicated. Would it be possible to write a generic function?:

```
def lookupOrSolve[F](tpe1: F, tpe2: F, renv: Set[Int])(implicit flix: Flix, alg: BoolAlg[F], cache: BoolCache[F]): ...
```

and then have:

```
if(typeVars.size > varThreshold) {
  lookupOrSolve(... pass in algebra and cache)
} else {
  lookupOrSolve(... pass in algebra and cache)
}
```

?